sklearn/datasets/value_dropper.py

```diff
@@ -0,0 +1,394 @@
+                                     " or an array of shape (n_features,). "
+                                     "%r was passed for class label %s"
+                                     % (val, class_key))
+
+        else:
+            missing_type = 'mcar'
+
+            # For MCAR
+            # Validate and convert the missing_distribution dict into a
+            # 1D probability distribution along the features
+
+            drop_probs = np.zeros((1, n_features), dtype=np.float64)
+
+            if isinstance(self.missing_distribution, (float, np.floating)):
+                drop_probs[:] = self.missing_distribution / n_features
+            elif isinstance(self.missing_distribution,
+                            (list, tuple, np.ndarray)):
+                missing_distribution = np.asarray(self.missing_distribution)
+                if missing_distribution.shape[0] != n_features:
+                    raise ValueError("The shape of the per feature "
+                                     "drop probabilities vector does not "
+                                     "conform to the number of features, %d"
+                                     % n_features)
+                drop_probs[:] = self.missing_distribution
+            else:
+                raise ValueError("missing_distribution must be a float or "
+                                 " 1D vector (list, tuple or np.ndarray) of "
+                                 "shape (n_features,) or dict of 1D vector / "
+                                 "floats. %r was passed"
+                                 % self.missing_distribution)
+
+        if 1 - drop_probs.ravel().sum() <= np.finfo(float).eps:
+            raise ValueError("The sum of all probabilities in the "
+                             "missing_distribution should sum up to less "
+                             "than 1. The sum was found to be %0.8f"
+                             % drop_probs.ravel().sum())
+
+        drop_counts = (drop_probs * n_values).astype(int)
+
+        if missing_type == 'mcar':
+            self._block_drop_missing_values(X, Ellipsis, 0, drop_counts, rng,
+                                            missing_type, missing_values)
+        else:
+            # For classification, NMAR, consider missing distribution
+            # based on class labels as subsets within data
+            for i, class_i in enumerate(classes):
+                self._block_drop_missing_values(X, y == class_i, i,
+                                                drop_counts, rng,
+                                                missing_type, missing_values)
+        return X
+
+    def _block_drop_missing_values(self, X, samples_mask, encoded_label,
+                                   drop_counts, rng, missing_type,
+                                   missing_values):
+        """Helper to insert missing values in given block (label)"""
+        n_features = X.shape[1]
+        this_block_indices = np.arange(X.shape[0])[samples_mask]
```

Comment:
>Which is why I felt it was both visually clearer and safer to pass the mask... WDYT?

You make a copy of X before and moreover X is modify on that function anyway. For me it's better to give `X[samples_mask]` as parameter and consider you just add NaN in a X on that function. It seems clearer no ?


> Also this particular line I had to do because samples_mask could be Ellipsis

On that case you just give the full X to the function.